File: be/test/util/filesystem_util_test.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "util/filesystem_util.h"

#include <gtest/gtest.h>
#include <sys/stat.h>

#include <filesystem>

#include "common/configbase.h"
#include "util/logging.h"

namespace doris {

namespace filesystem = std::filesystem;
using filesystem::path;

TEST(FileSystemUtil, rlimit) {
    ASSERT_LT(0ul, FileSystemUtil::max_num_file_handles());
}

TEST(FileSystemUtil, CreateDirectory) {
    char filename[] = "temp-XXXXXX";
    // Setup a temporary directory with one subdir
    std::string dir_name = mkdtemp(filename);
    path dir{dir_name};
    path subdir1 = dir / "path1";
    path subdir2 = dir / "path2";
    path subdir3 = dir / "a" / "longer" / "path";
    filesystem::create_directories(subdir1);
    // Test error cases by removing write permissions on root dir to prevent
    // creation/deletion of subdirs
    chmod(dir.string().c_str(), 0);
    if (getuid() == 0) { // User root
        EXPECT_TRUE(FileSystemUtil::create_directory(subdir1.string()).ok());
        EXPECT_TRUE(FileSystemUtil::create_directory(subdir2.string()).ok());
    } else { // User other
        EXPECT_FALSE(FileSystemUtil::create_directory(subdir1.string()).ok());
        EXPECT_FALSE(FileSystemUtil::create_directory(subdir2.string()).ok());
    }
    // Test success cases by adding write permissions back
    chmod(dir.string().c_str(), S_IRWXU);
    EXPECT_TRUE(FileSystemUtil::create_directory(subdir1.string()).ok());
    EXPECT_TRUE(FileSystemUtil::create_directory(subdir2.string()).ok());
    // Check that directories were created
    EXPECT_TRUE(filesystem::exists(subdir1) && filesystem::is_directory(subdir1));
    EXPECT_TRUE(filesystem::exists(subdir2) && filesystem::is_directory(subdir2));
    // Exercise VerifyIsDirectory
    EXPECT_TRUE(FileSystemUtil::verify_is_directory(subdir1.string()).ok());
    EXPECT_TRUE(FileSystemUtil::verify_is_directory(subdir2.string()).ok());
    EXPECT_FALSE(FileSystemUtil::verify_is_directory(subdir3.string()).ok());
    // Check that nested directories can be created
    EXPECT_TRUE(FileSystemUtil::create_directory(subdir3.string()).ok());
    EXPECT_TRUE(filesystem::exists(subdir3) && filesystem::is_directory(subdir3));
    // Cleanup
    filesystem::remove_all(dir);
}

TEST(FilesystemUtil, contain_path) {
    {
        std::string parent("/a/b");
        std::string sub("/a/b/c");
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }

    {
        std::string parent("/a/b/");
        std::string sub("/a/b/c/");
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }

    {
        std::string parent("/a///./././/./././b/"); // "/a/b/."
        std::string sub("/a/b/../././b/c/");        // "/a/b/c/"
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }

    {
        // relative path
        std::string parent("a/b/"); // "a/b/"
        std::string sub("a/b/c/");  // "a/b/c/"
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }
    {
        // relative path
        std::string parent("a////./././b/"); // "a/b/"
        std::string sub("a/b/../././b/c/");  // "a/b/c/"
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }
    {
        // absolute path and relative path
        std::string parent("/a////./././b/"); // "/a/b/"
        std::string sub("a/b/../././b/c/");   // "a/b/c/"
        EXPECT_FALSE(FileSystemUtil::contain_path(parent, sub));
        EXPECT_FALSE(FileSystemUtil::contain_path(sub, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(parent, parent));
        EXPECT_TRUE(FileSystemUtil::contain_path(sub, sub));
    }
}

} // end namespace doris

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: fe/fe-filesystem/fe-filesystem-api/src/test/java/org/apache/doris/filesystem/FileSystemUtilTest.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.filesystem;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;
import org.mockito.Mockito;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;

class FileSystemUtilTest {

    // --- extractParentDirectory ---

    @Test
    void extractParentDirectoryWithSlash() {
        Assertions.assertEquals("hdfs://nn/a/b/", FileSystemUtil.extractParentDirectory("hdfs://nn/a/b/file.csv"));
    }

    @Test
    void extractParentDirectoryRootFile() {
        Assertions.assertEquals("hdfs://nn/", FileSystemUtil.extractParentDirectory("hdfs://nn/file.csv"));
    }

    @Test
    void extractParentDirectoryNoSlash() {
        Assertions.assertEquals("file.csv", FileSystemUtil.extractParentDirectory("file.csv"));
    }

    @Test
    void extractParentDirectoryMultiLevel() {
        Assertions.assertEquals("s3://bucket/a/b/c/", FileSystemUtil.extractParentDirectory("s3://bucket/a/b/c/d.txt"));
    }

    @Test
    void extractParentDirectoryBucketRoot() {
        Assertions.assertEquals("s3://bucket/", FileSystemUtil.extractParentDirectory("s3://bucket/file"));
    }

    // --- asyncRenameFiles ---

    @Test
    void asyncRenameFilesCallsRenameForEachFile() throws Exception {
        FileSystem mockFs = Mockito.mock(FileSystem.class);
        List<CompletableFuture<?>> futures = new ArrayList<>();
        AtomicBoolean cancelled = new AtomicBoolean(false);
        List<String> fileNames = List.of("a.txt", "b.txt", "c.txt");

        FileSystemUtil.asyncRenameFiles(mockFs, Runnable::run, futures, cancelled,
                "s3://bucket/src/", "s3://bucket/dst/", fileNames);

        Assertions.assertEquals(3, futures.size());
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

        ArgumentCaptor<Location> srcCaptor = ArgumentCaptor.forClass(Location.class);
        ArgumentCaptor<Location> dstCaptor = ArgumentCaptor.forClass(Location.class);
        Mockito.verify(mockFs, Mockito.times(3)).rename(srcCaptor.capture(), dstCaptor.capture());

        List<Location> sources = srcCaptor.getAllValues();
        List<Location> dests = dstCaptor.getAllValues();
        Assertions.assertEquals(Location.of("s3://bucket/src/a.txt"), sources.get(0));
        Assertions.assertEquals(Location.of("s3://bucket/dst/a.txt"), dests.get(0));
        Assertions.assertEquals(Location.of("s3://bucket/src/b.txt"), sources.get(1));
        Assertions.assertEquals(Location.of("s3://bucket/dst/b.txt"), dests.get(1));
    }

    @Test
    void asyncRenameFilesSkipsWhenCancelled() throws Exception {
        FileSystem mockFs = Mockito.mock(FileSystem.class);
        List<CompletableFuture<?>> futures = new ArrayList<>();
        AtomicBoolean cancelled = new AtomicBoolean(true);

        FileSystemUtil.asyncRenameFiles(mockFs, Runnable::run, futures, cancelled,
                "s3://bucket/src", "s3://bucket/dst", List.of("a.txt"));

        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
        Mockito.verify(mockFs, Mockito.never()).rename(ArgumentMatchers.any(), ArgumentMatchers.any());
    }

    // --- asyncRenameDir ---

    @Test
    void asyncRenameDirCallsRenameDirectory() throws Exception {
        FileSystem mockFs = Mockito.mock(FileSystem.class);
        List<CompletableFuture<?>> futures = new ArrayList<>();
        AtomicBoolean cancelled = new AtomicBoolean(false);
        Runnable callback = Mockito.mock(Runnable.class);

        FileSystemUtil.asyncRenameDir(mockFs, Runnable::run, futures, cancelled,
                "s3://bucket/src", "s3://bucket/dst", callback);

        Assertions.assertEquals(1, futures.size());
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

        Mockito.verify(mockFs).renameDirectory(
                Location.of("s3://bucket/src"),
                Location.of("s3://bucket/dst"),
                callback);
    }

    @Test
    void asyncRenameDirSkipsWhenCancelled() throws Exception {
        FileSystem mockFs = Mockito.mock(FileSystem.class);
        List<CompletableFuture<?>> futures = new ArrayList<>();
        AtomicBoolean cancelled = new AtomicBoolean(true);

        FileSystemUtil.asyncRenameDir(mockFs, Runnable::run, futures, cancelled,
                "s3://bucket/src", "s3://bucket/dst", () -> {});

        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
        Mockito.verify(mockFs, Mockito.never()).renameDirectory(ArgumentMatchers.any(), ArgumentMatchers.any(), ArgumentMatchers.any());
    }
}
